### DISCUSSION TITLE
Registering commands in the GuildAvailable event doesn't seem to work

### DISCUSSION THREAD

**Discussion by Metritutus** [DSharpPlus/DSharpPlus#1632] [Support] [answered]
I'm attempting to register commands in the `GuildAvailable` event handler (which is how I did it with Discord.NET), however it doesn't seem to work.

I'm using the `DiscordShardedClient` approach.  If I do the following, my `test` command will be registered:

```csharp
            IReadOnlyDictionary<int, SlashCommandsExtension> slashCommands
                = await DiscordClient.UseSlashCommandsAsync(SlashCommandsConfiguration);

           slashCommands.RegisterCommands<TestSlashCommandModule>(<redacted>);
```

This however, will not register it:
```csharp
//...
          IReadOnlyDictionary<int, SlashCommandsExtension> slashCommands
              = await DiscordClient.UseSlashCommandsAsync(SlashCommandsConfiguration);

         DiscordClient.GuildAvailable += (sender, args) => DiscordClient_GuildAvailable(sender, args, slashCommands);
//...

        private async Task DiscordClient_GuildAvailable(DiscordClient sender,
                                                        GuildCreateEventArgs args,
                                                        IReadOnlyDictionary<int, SlashCommandsExtension> slashCommands)
        {
            slashCommands.RegisterCommands<TestCommandModule>(args.Guild.Id);
        }
```

My best guess is this relates to this somewhat vague paragraph in the [help documentation](https://dsharpplus.github.io/DSharpPlus/articles/slash_commands.html):

> It does not support registering or editing commands at runtime. While you can make commands at runtime using the methods on the client, if you have a command class registered for that guild/globally if you're making global commands, it will be overwritten (therefore probably deleted) on the next startup due to the limitations of the bulk overwrite endpoint. If your usage of slash commands depends on dynamically registering commands, this extension will not work for you.

The "at runtime" thing seems a bit odd; surely the app is running the whole time, but you'd need to register commands at some point?  Perhaps it refers to the `DiscordClient` being connected?  Would this be prior to `StartAsync()` being called for the `DiscordShardClient`?  If so, how are you supposed to get the guild ids to register the commands with without knowing them in advance and hard-coding them?  If the `GuildAvailable` event happens too late for slash command registration, how do you register slash commands, when starting the bot, only for the guilds the bot is in (or specific guilds filtered based on some criteria) without hard-coding the guild ids or using global commands?

I don't need to dynamically change the commands while the bot is running, but I would like to set non-global slash commands on startup.

What am I missing?

**Comment by Plerx2493** [comment, marked as answer]
You have to register your commands before connecting to the gateway. Like you said before `ConnectAsync`. If you want to lock your commands to one specific guild, you should save the id in your config or hardcode it. If you want to register them for all guild then simply dont provide a id and they will be registered globally

**Reply by Metritutus** [reply]
Ah, darn, that is unfortunate.  I'm guessing this is an implementation detail specific to DSharpPlus, as Discord.NET didn't seem to have this limitation?

**Reply by akiraveliara** [reply]
it is an impl detail, but fwiw you can register your commands later and then refresh them

**Reply by Metritutus** [reply]
Ah, how does one go about refreshing slash commands if they've been registered later, other than by restarting the bot?

**Reply by akiraveliara** [reply]
calling the refresh function on the extension

**Reply by Metritutus** [reply]
Sorry for the delayed response.  Thanks for your responses, cheers!
